app/controllers/sessions_controller.rb

```diff
@@ -1,13 +1,24 @@
 class SessionsController < ApplicationController
   def new
+
   end
 
   def create
-    if user = User.from_omniauth(request.env["omniauth.auth"])
-      session[:user_id] = user.id
-      redirect_to root_path
+    if request.env["omniauth.auth"]
+      if user = User.from_omniauth(request.env["omniauth.auth"])
+        session[:user_id] = user.id
+        redirect_to root_path
+      else
+        redirect_to root_path
+      end
     else
-      redirect_to root_path
+      user = User.find_by(email: params[:session]["email"])
+      if user && user.authenticate(params[:session]["password"])
+        session[:user_id] = user.id
+        redirect_to root_path
+      else
+        redirect_to login_path
+      end
```

Comment:
Same comments as above, nicely done. For potential refactors do we want to do something like this.
```
def create
   if request.env["omniauth.auth"]
      validate_oauth_user
   else
      validate_manual_user
   end
end

private

def validate_oauth_user
    if user = User.from_omniauth(request.env["omniauth.auth"])
      session[:user_id] = user.id
      redirect_to root_path
    else
      redirect_to root_path
    end
end

def validate_manual_user
    user = User.find_by(email: params[:session]["email"])
    if user && user.authenticate(params[:session]["password"])
        session[:user_id] = user.id
        redirect_to root_path
    else
       redirect_to login_path
    end
end
```

ultimately don't we want to build a `current_user` helper method?

Reply:
Maybe refactor in the model. I don't think we should add non restful actions in the controller